### PR TITLE
Change cumulative compaction for decoupling storage from compution

### PR DESCRIPTION
**Pull request by chaoyli**
1. Calculate cumulative point when loading tablet first time.
2. Simplify pick rowsets logic upon delete predicate.
3. Saving meta and modify rowsets only once after cumulative compaction.

### CODE REVIEW

File: be/src/olap/base_compaction.h [hunk 1]
```diff
@@ -32,152 +33,45 @@ namespace doris {
...
-    // - 其它情况下，返回相应的错误码
-    OLAPStatus init(TabletSharedPtr tablet, bool is_manual_trigger = false);
-
-    // 执行BaseCompaction, 可能会持续很长时间
-    //
-    // 返回值：
-    // - 如果执行成功，则返回OLAP_SUCCESS；
-    // - 其它情况下，返回相应的错误码
-    OLAPStatus run();
+    BaseCompaction(TabletSharedPtr tablet);
```
Comment (imay) [inline be/src/olap/base_compaction.h]: you should add comments for these public functions. It's better to add an example for how to use this class.

If some functions aren't intended to be used by others, please make them private

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/olap/base_compaction.cpp [hunk 2]
```diff
@@ -40,487 +39,264 @@ using std::vector;
 
 namespace doris {
```
Comment (yiguolei) [inline be/src/olap/base_compaction.cpp]: It's better to move the rowset_comparator to rowset.h file

File: be/src/olap/base_compaction.cpp [hunk 3]
```diff
@@ -40,487 +39,264 @@ using std::vector;
...
+      _input_row_num(0)
+    { }
 
-    // 1. 尝试取得base compaction的锁
-    if (!_try_base_compaction_lock()) {
-        LOG(WARNING) << "another base compaction is running. tablet=" << tablet->full_name();
-        return OLAP_ERR_BE_TRY_BE_LOCK_ERROR;
+BaseCompaction::~BaseCompaction() {
+    if (_base_locked) {
+        _tablet->release_base_compaction_lock();
```
Comment (yiguolei) [inline be/src/olap/base_compaction.cpp]: It's too trick. Add a scoped lock or mutex.

Comment (yiguolei) [inline be/src/olap/base_compaction.cpp]: It's too trick to add a variable _base_locked to indicate whether base compaction lock is owned. It's better to add a scoped mutex like migration lock.

File: be/src/olap/base_compaction.cpp [hunk 4]
```diff
@@ -40,487 +39,264 @@ using std::vector;
...
-    {
-        DorisMetrics::base_compaction_deltas_total.increment(_need_merged_versions.size());
-        int64_t merge_bytes = 0;
-        for (auto& rowset : rowsets) {
-            merge_bytes += rowset->data_disk_size();
-        }
-        DorisMetrics::base_compaction_bytes_total.increment(merge_bytes);
+OLAPStatus BaseCompaction::pick_rowsets_to_compact() {
+    _input_rowsets.clear();
+    _tablet->pick_candicate_rowsets_to_base_compaction(&_input_rowsets);
```
Comment (yiguolei) [inline be/src/olap/base_compaction.cpp]: should use capture rs readers to check data is valid

File: be/src/olap/base_compaction.cpp [hunk 5]
```diff
@@ -40,487 +39,264 @@ using std::vector;
...
 
+OLAPStatus BaseCompaction::save_meta() {
     return OLAP_SUCCESS;
 }
 
-OLAPStatus BaseCompaction::_update_header(const vector<RowsetSharedPtr>& unused_rowsets) {
-    WriteLock wrlock(_tablet->get_header_lock_ptr());
+OLAPStatus BaseCompaction::modify_rowsets() {
+    std::vector<RowsetSharedPtr> output_rowsets;
+    output_rowsets.push_back(_output_rowset);
```
Comment (yiguolei) [inline be/src/olap/base_compaction.cpp]: Need tablet write lock？

File: be/src/olap/cumulative_compaction.cpp [hunk 6]
```diff
@@ -32,356 +32,121 @@ using std::vector;
...
-    _tablet->release_header_lock();
-    // 如果为-1，则该table之前没有设置过cumulative layer point
-    // 我们在这里设置一下
-    if (_old_cumulative_layer_point == -1) {
-        LOG(INFO) << "tablet has an unreasonable cumulative layer point. tablet=" << _tablet->full_name()
-                  << ", cumulative_layer_point=" << _old_cumulative_layer_point;
-        _tablet->release_cumulative_lock();
-        return OLAP_ERR_CUMULATIVE_INVALID_PARAMETERS;
-    }
+    _cumulative_locked = true;
```
Comment (yiguolei) [inline be/src/olap/cumulative_compaction.cpp]: Use scoped lock here

File: be/src/olap/cumulative_compaction.cpp [hunk 7]
```diff
@@ -32,356 +32,121 @@ using std::vector;
...
     }
-
-    if (delta_versions->size() == 0) {
-        LOG(INFO) << "no delta versions. cumulative_point=" << _old_cumulative_layer_point;
-        return OLAP_ERR_CUMULATIVE_NO_SUITABLE_VERSIONS;
+		
+    if (_input_rowsets.empty()) {
+        _tablet->set_cumulative_layer_point(candidate_rowsets.back()->end_version() + 1);
+    } else {
+        _tablet->set_cumulative_layer_point(_input_rowsets.back()->end_version() + 1);
```
Comment (yiguolei) [inline be/src/olap/cumulative_compaction.cpp]: Must not update ce point here. Should update it after ce success or ce failed

File: be/src/olap/tablet.cpp [hunk 8]
```diff
@@ -736,6 +749,32 @@ OLAPStatus Tablet::max_continuous_version_from_begining(Version* version, Versio
     return OLAP_SUCCESS;
 }
 
+OLAPStatus Tablet::calcuate_cumulative_point() {
```
Comment (yiguolei) [inline be/src/olap/tablet.cpp]: it is not used out of tablet. set it to private

File: be/src/olap/tablet.cpp [hunk 9]
```diff
@@ -736,6 +749,32 @@ OLAPStatus Tablet::max_continuous_version_from_begining(Version* version, Versio
...
+    ReadLock rdlock(&_meta_lock);
+    std::list<Version> existing_versions;
```
Comment (yiguolei) [inline be/src/olap/tablet.cpp]: It is used during init. Does it need meta lock really?

File: be/src/olap/tablet.cpp [hunk 10]
```diff
@@ -933,4 +972,20 @@ OLAPStatus Tablet::set_partition_id(int64_t partition_id) {
     return _tablet_meta->set_partition_id(partition_id);
 }
 
+void Tablet::pick_candicate_rowsets_to_cumulative_compaction(std::vector<RowsetSharedPtr>* candidate_rowsets) {
+    for (auto& it : _rs_version_map) {
+        if (it.first.first > _cumulative_point) {
+            candidate_rowsets->push_back(it.second);
```
Comment (yiguolei) [inline be/src/olap/tablet.cpp]: add first == second here

File: be/src/olap/base_compaction.cpp [hunk 11]
```diff
@@ -40,487 +39,264 @@ using std::vector;
...
-OLAPStatus BaseCompaction::init(TabletSharedPtr tablet, bool is_manual_trigger) {
-    // 表在首次查询或PUSH等操作时，会被加载到内存
-    // 如果表没有被加载，表明该表上目前没有任何操作，所以不进行BE操作
-    if (!tablet->init_succeeded()) {
-        return OLAP_ERR_INPUT_PARAMETER_ERROR;
-    }
-
-    LOG(INFO) << "init base compaction handler. [tablet=" << tablet->full_name() << "]";
+static bool rowset_comparator(const RowsetSharedPtr& left, const RowsetSharedPtr& right) {
+    return left->end_version() < right->end_version();
```
Comment (yiguolei) [inline be/src/olap/base_compaction.cpp]: Should add a parent class and move all common methods and methods only used by compaction to parent class. There are too many duplicate methods.

**Review by yiguolei — COMMENTED** [review]
some comments

File: be/src/olap/base_compaction.cpp (hunk 3 above)
Comment (chaoyli) [reply be/src/olap/base_compaction.cpp]: OK, I will change it to another form

File: be/src/olap/tablet.cpp [hunk 12]
```diff
@@ -736,6 +749,31 @@ OLAPStatus Tablet::max_continuous_version_from_begining(Version* version, Versio
...
+
+    // sort the existing versions in ascending order
+    existing_versions.sort([](const Version& a, const Version& b) {
+        // simple because 2 versions are certainly not overlapping
+        return a.first < b.first;
+    });
+
+    _cumulative_point = -1;
+    int64_t prev_version = -1;
+    for (const Version& version : existing_versions) {
```
Comment (yiguolei) [inline be/src/olap/tablet.cpp]: prev_version not updated?

**Review by yiguolei — DISMISSED** [review]
LGTM

File: be/src/olap/tablet_manager.cpp [hunk 13]
```diff
@@ -737,18 +737,16 @@ TabletSharedPtr TabletManager::find_best_tablet_to_compaction(
...
                     continue;
-                } else {
-                    table_ptr->release_cumulative_lock();
                 }
             }
 
             if (compaction_type == CompactionType::BASE_COMPACTION) {
-                if (!table_ptr->try_base_compaction_lock()) {
+                MutexLock lock(table_ptr->get_base_lock());
+                if (!lock.own_lock()) {
```
Comment (imay) [inline be/src/olap/tablet_manager.cpp]: This is not a try-lock, what does this check mean?

File: be/src/olap/base_compaction.cpp [hunk 14]
```diff
@@ -16,511 +16,145 @@
...
-    if (base_compaction_layer_point == -1) {
-        VLOG(3) << "can't do base compaction: no cumulative files."
-                << "tablet=" << _tablet->full_name()
-                << ", base_version=0-" << _old_base_version.second
-                << ", cumulative_layer_point=" << cumulative_layer_point + 1;
-        return false;
+OLAPStatus BaseCompaction::pick_rowsets_to_compact() {
+    _input_rowsets.clear();
+    _tablet->pick_candicate_rowsets_to_base_compaction(&_input_rowsets);
+    if (_input_rowsets.size() == 0 || _input_rowsets.size() == 1) {
```
Comment (imay) [inline be/src/olap/base_compaction.cpp]:
```suggestion
    if (_input_rowsets.size() <= 1) {
```

File: be/src/olap/tablet.cpp [hunk 15]
```diff
@@ -933,4 +973,21 @@ OLAPStatus Tablet::set_partition_id(int64_t partition_id) {
...
+    for (auto& it : _rs_version_map) {
+        if (it.first.first >= _cumulative_point
+            && it.first.first == it.first.second) {
+            candidate_rowsets->push_back(it.second);
+        }
+    }
+}
+
+void Tablet::pick_candicate_rowsets_to_base_compaction(std::vector<RowsetSharedPtr>* candidate_rowsets) {
+    for (auto& it : _rs_version_map) {
```
Comment (imay) [inline be/src/olap/tablet.cpp]: Is this thread-safe?

File: be/src/olap/storage_engine.cpp [hunk 16]
```diff
@@ -598,27 +585,13 @@ void StorageEngine::perform_base_compaction(DataDir* data_dir) {
     TabletSharedPtr best_tablet = _tablet_manager->find_best_tablet_to_compaction(CompactionType::BASE_COMPACTION, data_dir);
     if (best_tablet == nullptr) { return; }
 
-    DorisMetrics::base_compaction_request_total.increment(1);
```
Comment (imay) [inline be/src/olap/storage_engine.cpp]: why do you drop this metrics?

File: be/src/olap/tablet_manager.cpp (hunk 13 above)
Comment (chaoyli) [reply be/src/olap/tablet_manager.cpp]: I missed it. I will add it in next commit.

File: be/src/olap/storage_engine.cpp (hunk 16 above)
Comment (chaoyli) [reply be/src/olap/storage_engine.cpp]: I drop it wrong. I add it back.

File: be/src/olap/tablet.cpp (hunk 15 above)
Comment (chaoyli) [reply be/src/olap/tablet.cpp]: I will add read lock to it.

File: be/src/olap/tablet_manager.cpp [hunk 17]
```diff
@@ -301,6 +301,11 @@ OLAPStatus TabletManager::create_tablet(const TCreateTabletReq& request,
         LOG(WARNING) << "fail to create tablet. res=" << res;
     }
 
+    // call calculate_cumulative_point() manually.
+    // If not, cumulative_point will set to minus one
+    // when init tablet during create tablet.
+    RETURN_NOT_OK(tablet->calculate_cumulative_point());
+
```
Comment (yiguolei) [inline be/src/olap/tablet_manager.cpp]: If calculate method is called in two cases, it's better to call it during cumulative compaction.

**Review by yiguolei — COMMENTED** [review]
some comments

**Review by yiguolei — DISMISSED** [review]
LGTM

File: be/src/olap/base_compaction.cpp [hunk 18]
```diff
@@ -16,511 +16,151 @@
...
-    if (OLAP_SUCCESS != _tablet->capture_consistent_versions(_new_base_version, candidate_versions)) {
-        LOG(WARNING) << "fail to select shortest version path."
-            << "start=" << _new_base_version.first
-            << ", end=" << _new_base_version.second;
-        return  false;
+OLAPStatus BaseCompaction::pick_rowsets_to_compact() {
+    _input_rowsets.clear();
+    _tablet->pick_candicate_rowsets_to_base_compaction(&_input_rowsets);
+    if (_input_rowsets.size() <= 1) {
+        LOG(INFO) << "There is no enough rowsets to cumulative compaction."
```
Comment (imay) [inline be/src/olap/base_compaction.cpp]: Do we need this here as INFO?

File: be/src/olap/base_compaction.cpp [hunk 19]
```diff
@@ -16,511 +16,151 @@
...
-            return OLAP_ERR_ROWSET_CREATE_READER;
-        }
-        rs_readers.push_back(rs_reader);
+OLAPStatus BaseCompaction::do_compaction() {
+    DorisMetrics::base_compaction_deltas_total.increment(_input_rowsets.size());
+    for (auto& rowset : _input_rowsets) {
+        _input_rowsets_size += rowset->data_disk_size();
+        _input_row_num += rowset->num_rows();
     }
+    DorisMetrics::base_compaction_bytes_total.increment(_input_rowsets_size);
```
Comment (imay) [inline be/src/olap/base_compaction.cpp]: I think we should update metric when compaction is finished

And this metric can be get from compaction process

File: be/src/olap/base_compaction.cpp [hunk 20]
```diff
@@ -16,511 +16,151 @@
...
-    //       无法达到删除数据的目的
-    // 想法：如果一定要使用MassiveMerger，这里可以提供一种方案
-    //       1. 在此处加一个检查，检测此次BE是否包含删除条件, 即检查Reader中
-    //          ReaderParams的delete_handler
-    //       2. 如果包含删除条件，则不使用MassiveMerger，使用Merger
-    //       3. 如果不包含删除条件，则可以使用MassiveMerger
-    uint64_t merged_rows = 0;
-    uint64_t filted_rows = 0;
-    OLAPStatus res = OLAP_SUCCESS;
+    RETURN_NOT_OK(construct_output_rowset_writer());
```
Comment (imay) [inline be/src/olap/base_compaction.cpp]: I think it's better to encapsulate all the following logic into base class.

File: be/src/olap/base_compaction.cpp [hunk 21]
```diff
@@ -16,511 +16,151 @@
...
-    OlapStopWatch watch;
-    vector<RowsetReaderSharedPtr> rs_readers;
-    for (auto& rowset : rowsets) {
-        RowsetReaderSharedPtr rs_reader(rowset->create_reader());
-        if (rs_reader == nullptr) {
-            LOG(WARNING) << "rowset create reader failed. rowset:" <<  rowset->rowset_id();
-            return OLAP_ERR_ROWSET_CREATE_READER;
-        }
-        rs_readers.push_back(rs_reader);
+OLAPStatus BaseCompaction::do_compaction() {
```
Comment (imay) [inline be/src/olap/base_compaction.cpp]: I think it's better to encapsulate all the following logic into base class.

Comment (chaoyli) [reply be/src/olap/base_compaction.cpp]: I do not think so.
Different Compaction can has different logic. 
If I want to change base compaction to capture partial rowsets, I can do it in this function.

Comment (imay) [reply be/src/olap/base_compaction.cpp]: most of this function's logic is the same with cumulative compaction. 
1. pick rowset
2. check rowset if continue
3. create readers and writer
4. check rows if match
5. write metadata

Now, there are two same logic in two files, I think we can make it better.

File: be/src/olap/base_compaction.cpp (hunk 18 above)
Comment (chaoyli) [reply be/src/olap/base_compaction.cpp]: I will change it to LOG(WARNING)

Comment (imay) [reply be/src/olap/base_compaction.cpp]: I think we should remove this LOG.

File: be/src/olap/base_compaction.cpp [hunk 22]
```diff
@@ -16,513 +16,115 @@
...
-        LOG(INFO) << "acquire data sources failed. version="
-           << spec_version.first << "-" << spec_version.second;
-        return OLAP_ERR_BE_ERROR_DELETE_ACTION;
-    }
-
-    VLOG(3) << "delete file action is OK";
-
-    return OLAP_SUCCESS;
+ReaderType BaseCompaction::compaction_type() const {
+    return _compaction_type;
```
Comment (imay) [inline be/src/olap/base_compaction.cpp]: there is no need to store it in class attribute, It's OK to just return a constant value.

File: be/src/olap/compaction.h [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+// Any compaction should go through four procedures.
+//  1. pick rowsets satisfied to compact
+//  2. do compaction
+//  3. modify rowsets
+//  4. gc unused rowstes
+class Compaction {
+public:
+    Compaction(TabletSharedPtr tablet);
+    virtual ~Compaction();
+
```
Comment (imay) [inline be/src/olap/compaction.h]: Only keep needed function as public, make others private/protected

File: be/src/olap/base_compaction.cpp [hunk 24]
```diff
@@ -16,513 +16,115 @@
...
-                     << "-" << _new_base_version.second
-                     << ", vector_version=" << candidate_versions.begin()->first
-                     << "-" << candidate_versions.rbegin()->second;
-        return false;
-    }
-
-    VLOG(10) << "valid need merged version";
-    return true;
+const std::string& BaseCompaction::compaction_name() const {
+    return _compaction_name;
```
Comment (imay) [inline be/src/olap/base_compaction.cpp]: Just return const value too

Comment (chaoyli) [reply be/src/olap/base_compaction.cpp]: I want to use reference to reduce string copy.
If use const value, it will cause string copy.

**Review by yiguolei — COMMENTED** [review]
LGTM

File: be/src/olap/base_compaction.cpp [hunk 25]
```diff
@@ -16,513 +16,114 @@
...
-    if (rs_readers.empty()) {
-        LOG(INFO) << "acquire data sources failed. version="
-           << spec_version.first << "-" << spec_version.second;
-        return OLAP_ERR_BE_ERROR_DELETE_ACTION;
-    }
-
-    VLOG(3) << "delete file action is OK";
-
-    return OLAP_SUCCESS;
+ReaderType BaseCompaction::compaction_type() const {
```
Comment (imay) [inline be/src/olap/base_compaction.cpp]: put this to .h file

File: be/src/olap/base_compaction.h [hunk 26]
```diff
@@ -18,166 +18,30 @@
...
-    Version _old_base_version;
-    // 现有的版本号最大的cumulative
-    Version _latest_cumulative;
-    // 在此次base compaction执行过程中，将被合并的cumulative文件版本
-    std::vector<Version> _need_merged_versions;
-    // 需要新增的版本对应Rowset的
-    std::vector<RowsetSharedPtr> _new_rowsets;
-    bool _base_compaction_locked;
-    RowsetWriterSharedPtr _rs_writer;
+    std::string _compaction_name;
```
Comment (imay) [inline be/src/olap/base_compaction.h]: delete these

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**